eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php

```diff
@@ -1160,4 +1175,89 @@ public function bulkRemoveTranslation($languageId, $actions)
         ;
         $query->execute();
     }
+
+    /**
+     * Archive (remove or historize) URL aliases for removed Translations.
+     *
+     * @param int $locationId
+     * @param int $parentId
+     * @param int[] $languageIds Language IDs of removed Translations
+     */
+    public function archiveUrlAliasesForDeletedTranslations($locationId, $parentId, array $languageIds)
+    {
+        // determine proper parent for linking historized entry
+        $existingLocationEntry = $this->loadAutogeneratedEntry(
+            'eznode:' . $locationId,
+            $parentId
+        );
+
+        // filter existing URL alias entries by any of the specified removed languages
+        $rows = $this->loadLocationEntriesMatchingMultipleLanguages(
+            $locationId,
+            $languageIds
+        );
+
+        // remove specific languages from a bit mask
+        foreach ($rows as $row) {
+            // filter mask to reduce the number of calls to storage engine
+            $rowLanguageMask = (int) $row['lang_mask'];
+            $languageIdsToBeRemoved = array_filter(
+                $languageIds,
+                function ($languageId) use ($rowLanguageMask) {
+                    return $languageId & $rowLanguageMask;
+                }
+            );
+
+            if (empty($languageIdsToBeRemoved)) {
+                continue;
+            }
+
+            // use existing entry to link archived alias or use current alias id
+            $linkToId = !empty($existingLocationEntry) ? $existingLocationEntry['id'] : $row['id'];
+            foreach ($languageIdsToBeRemoved as $languageId) {
+                $this->archiveUrlAliasForDeletedTranslation(
+                    $row['lang_mask'],
+                    $languageId,
+                    $row['parent'],
+                    $row['text_md5'],
+                    $linkToId
+                );
+            }
+        }
+    }
+
+    /**
+     * Load list of aliases for given $locationId matching any of the Languages specified by $languageMask.
+     *
+     * @param int $locationId
+     * @param int[] $languageIds
+     *
+     * @return \Generator
```

Comment:
what's the actually type that should be type hinted here?

Reply:
It is an array of associative database row arrays, however what this method actually returns (from a consumer POV) is an instance of `\Generator`, so I'm not sure if hinting it in any other way would be correct. What did you have in mind?

Reply:
Imho returning generator is a internal detail, php doc should rather focus on what the end user is going to consume here. Examples:
https://docs.phpdoc.org/guides/types.html#arrays

If it's important to make it clear it generator, on intneral classes  (not api/spi interfaces) it could be hinted as `array[]|\Generator`

Reply:
Ok, `array[]|\Generator` works for me. Fixup: fb6222a